pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -941,6 +955,364 @@ class StateVectorCudaMPI
+                              .csrOffsets.size();
+                     i0++) {
+                    splitSparseMatrix[block_row_id][block_col_id]
+                        .csrOffsets[i0] +=
+                        splitSparseMatrix[block_row_id][block_col_id]
+                            .csrOffsets[i0 - 1];
+                }
+            }
+        }
+
+        return splitSparseMatrix;
+    }
+
+    /**
+     * @brief expval(H) calculation with cuSparseSpMV.
+     *
+     * @tparam index_type Integer type used as indices of the sparse matrix.
+     * @param csr_Offsets_ptr Pointer to the array of row offsets of the sparse
+     * matrix. Array of size csrOffsets_size.
+     * @param csrOffsets_size Number of Row offsets of the sparse matrix.
+     * @param columns_ptr Pointer to the array of column indices of the sparse
+     * matrix. Array of size numNNZ
+     * @param values_ptr Pointer to the array of the non-zero elements
+     * @param numNNZ Number of non-zero elements.
+     * @return auto Expectation value.
+     */
+    template <class index_type>
+    auto getExpectationValueOnSparseSpMV(
+        const index_type *csrOffsets_ptr, const index_type csrOffsets_size,
+        const index_type *columns_ptr,
+        const std::complex<Precision> *values_ptr, const index_type numNNZ) {
+        if (mpi_manager_.getRank() == 0) {
+            PL_ABORT_IF_NOT(static_cast<size_t>(csrOffsets_size - 1) ==
+                                (size_t{1} << this->getTotalNumQubits()),
+                            "Incorrect size of CSR Offsets.");
+            PL_ABORT_IF_NOT(numNNZ > 0, "Empty CSR matrix.");
+        }
+
+        const CFP_t alpha = {1.0, 0.0};
+        const CFP_t beta = {0.0, 0.0};
+
+        Precision local_expect = 0;
+
+        auto device_id = BaseType::getDataBuffer().getDevTag().getDeviceID();
+        auto stream_id = BaseType::getDataBuffer().getDevTag().getStreamID();
+
+        cudaDataType_t data_type;
+        cusparseIndexType_t compute_type;
+
+        if constexpr (std::is_same_v<CFP_t, cuDoubleComplex> ||
+                      std::is_same_v<CFP_t, double2>) {
+            data_type = CUDA_C_64F;
+            compute_type = CUSPARSE_INDEX_64I;
+        } else {
+            data_type = CUDA_C_32F;
+            compute_type = CUSPARSE_INDEX_32I;
+        }
+
+        std::vector<std::vector<CSRMatrix<Precision, index_type>>>
+            csrmatrix_blocks;
+
+        size_t num_col_blocks = mpi_manager_.getSize();
+        size_t num_row_blocks = mpi_manager_.getSize();
+
+        if (mpi_manager_.getRank() == 0) {
+            csrmatrix_blocks =
+                splitCSRMatrix(num_col_blocks, num_row_blocks, csrOffsets_ptr,
+                               columns_ptr, values_ptr);
+        }
+
+        const size_t length_local = size_t{1} << this->getNumLocalQubits();
+
+        DataBuffer<CFP_t, int> d_tmp{length_local, device_id, stream_id, true};
```

Comment:
What data do `d_tmp` and `d_tmp_res` hold? I wonder if we can use something other than `tmp` to explain intent.

Reply:
`d_tmp` is the result per block and `d_tmp_res` is the result of a row of block.